### ISSUE TITLE
Hyper Params & loss function for ViT-L/16

### ISSUE THREAD

**Issue by owmohamm** [google-research/vision_transformer#34]
Hello!
I was trying to pretrain the ViT-L/16 architecture on imagenet-21k though there are a few aspects which I'm unsure of and hoping you can clarify.

1. What were the hyperparameters used for pretraining. In the paper it's mentioned that the model was trained for either 30/90 epochs and 0.1 weight decay at the start and 0.03 in the appendix, hence the uncertainty.

2. Is the data augmentation and model initialization the same as what's used in the released jax code?

3. What loss function was used? considering that imagenet-21k and JFT-300M are multilabel datasets.

Thanks!

**Comment by andsteing** [comment]
Hello!

1. The checkpoints published in this repository were pretrained for 90 epochs.
2. Same data augmentation and model initialization
3. We use sigmoid cross entropy (see e.g. [tf.nn.softmax_cross_entropy_with_logits](https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits)) - you could implement this as follows:

```python
def sigmoid_xent(*, logits, labels):
  log_p = jax.nn.log_sigmoid(logits)
  log_not_p = jax.nn.log_sigmoid(-logits)
  nll = -jnp.sum(labels * log_p + (1. - labels) * log_not_p, axis=-1)
  return jnp.mean(nll)
```

Best!

**Comment by owmohamm** [comment]
Thanks for the info!!

Just to confirm, you were referring to [tf.nn.sigmoid_cross_entropy_with_logits](https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits) and not [tf.nn.softmax_cross_entropy_with_logits](https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits), right?

In regards to the weight decay, what value was used and was it applied to full model or were the bias and norm terms excluded? When I try to pretrain the model the bias terms tend to grow much larger than those in the released weights

**Comment by andsteing** [comment]
You're 100% right, that was an unfortunate typo on my side!

As for weight decay, we used `weight_decay=0.1` and we apply the following function to all `kernel` weights:

```python
def decay_weight_fn(w, lr, weight_decay):
  return (1.0 - lr * decay) * w if decay else w
```

As for bias terms, we initialized the head biases to -10 :

```
params["head"]["bias"] = jnp.full_like(params["head"]["bias"], -10.)
```

HTH

**Comment by owmohamm** [comment]
Thanks for your help!!, that resolves any questions I had.

**Comment by andsteing** [comment]
Correction : Actually, for the relevant pretraining run, we used `flax.optim.Adam(weight_decay=0.1)` instead of the manual version that I described above (so this would indeed apply to the weight decay to all the weights).

**Comment by owmohamm** [comment]
That's interesting, was this done to combat overfitting?
In my experiments (not this architecture but in general) I've noticed that adding weight decay for the bias terms usually gives an inferior model, but in this case i.e. pretraining ViT-L/16 on imagenet-21k the model tends to overfit quite easily.

**Comment by akolesnikoff** [comment]
Hi @owmohamm  and sorry for the late reply.

AFAIK weight decay does not have significant regularization effect in the modern networks that have normalization layers (e.g. LayerNorm) or standardized weights. I can recommend to look into this two references to understand this effect:
1. L2 Regularization versus Batch and Weight Normalization by Laarhoven: https://arxiv.org/abs/1706.05350.
2. An Exponential Learning Rate Schedule for Deep Learning by Li and Arora - https://arxiv.org/abs/1910.07454.

Hope it helps.

**Comment by pengzhiliang** [comment]
Ok, thank u very much!
